src/osm4routing/reader.rs

```diff
@@ -1,259 +1,434 @@
-        for way in &self.ways {
-            for (i, node_id) in way.nodes.iter().enumerate() {
-                let node = self
-                    .nodes
-                    .get_mut(node_id)
-                    .expect("Missing node, id: {node_id}");
-                // Count double extremities nodes
-                if i == 0 || i == way.nodes.len() - 1 {
-                    node.uses += 2;
-                } else {
-                    node.uses += 1;
-                }
-            }
-        }
+  /// Performs sanity checks on nodes and ways for debugging purposes.
+  ///
+  /// This function checks for missing nodes in ways and removes the ways with missing nodes.
+  /// It also prints statistics about the removed ways and missing nodes to aid in debugging.
+  /// This function is intended for debugging purposes and should not be used in production code.
+  fn sanity_checks_nodes(&mut self) {
+    let mut ways_to_remove = Vec::new();
+    let mut removed_ways_count = 0;
+    let mut missing_node_stats = Vec::new();
+
+    for (way_index, way) in self.ways.iter().enumerate() {
+      let mut missing_node = false;
+
+      for (i, node_id) in way.nodes.iter().enumerate() {
+          if let Some(node) = self.nodes.get_mut(node_id) {
+              // Count double extremities nodes
+              if i == 0 || i == way.nodes.len() - 1 {
+                  node.uses += 2;
+              } else {
+                  node.uses += 1;
+              }
+          } else {
+              missing_node = true;
+              missing_node_stats.push((way_index, *node_id));
+              break;
+          }
+      }
+
+      if missing_node {
+          ways_to_remove.push(way_index);
+          removed_ways_count += 1;
+      }
     }
 
-    fn split_way(&self, way: &Way) -> Vec<Edge> {
-        let mut result = Vec::new();
-
-        let mut source = NodeId(0);
-        let mut geometry = Vec::new();
-        let mut nodes = Vec::new();
-        for (i, &node_id) in way.nodes.iter().enumerate() {
-            let node = self.nodes[&node_id];
-            geometry.push(node.coord);
-            nodes.push(node.id);
-            if i == 0 {
-                source = node_id;
-            } else if node.uses > 1 {
-                result.push(Edge {
-                    id: format!("{}-{}", way.id.0, result.len()),
-                    osm_id: way.id,
-                    source,
-                    target: node_id,
-                    geometry,
-                    properties: way.properties,
-                    nodes,
-                });
-
-                source = node_id;
-                geometry = vec![node.coord];
-                nodes = vec![node.id]
-            }
-        }
-        result
+    let removed_ways_percentage = (removed_ways_count as f32 / self.ways.len() as f32) * 100.0;
+
+    // remove ways with missing nodes
+    for way_index in ways_to_remove.iter().rev() {
+        self.ways.remove(*way_index);
```

Comment:
An optionnal verbose output could be useful here at some point.

Reply:
we only manage to have elements to remove if the file is not properly formatted (which is also why there are print lines afterwards). If the input files are correct, we'll never enter this loop (the purpose of the project). It was still having problems with the contents of the files when writing, this loop was important to remove the elements causing errors (This is why this function is a sanity check).

Reply:
we only manage to have elements to remove if the file is not properly formatted (which is also why there are print lines afterwards). If the input files are correct, we'll never enter this loop (the purpose of the project). It was still having problems with the contents of the files when writing, this loop was important to remove the elements causing errors (This is why this function is a sanity check).